src/sync/spsc_bufferqueue.rs

```diff
@@ -0,0 +1,398 @@
+//! SPSC ringbuffer
+
+use std::sync::atomic::Ordering::{Acquire, Release, Relaxed};
+use std::sync::atomic::{AtomicUsize, AtomicBool};
+use std::sync::Arc;
+use std::ptr;
+use std::mem;
+use std::cmp;
+use std::marker::PhantomData;
+use mem::CachePadded;
+
+#[inline(always)]
+unsafe fn deallocate<T>(ptr: *mut T, old_size: usize) {
+    Vec::from_raw_parts(ptr, 0, old_size);
+}
+
+#[inline(always)]
+unsafe fn allocate<T>(size: usize) -> *mut T {
+    let mut buf = Vec::with_capacity(size);
+    let ptr = buf.as_mut_ptr();
+    mem::forget(buf);
+    ptr
+}
+
+/// A single-producer, single consumer bounded wait-free ringbuffer queue
+///
+/// All operations on the buffer queue are wait-free,
+/// provided move/copy operations are waitfree.
+/// This queue does not allocate after constructions
+#[repr(C)]
+pub struct SpscBufferQueue<T: Send> {
+    // This is a pointer instead of a vector
+    // so that vector doesn't call constructors
+    data_block: *mut T,
+    size: usize,
+    _marker: PhantomData<T>,
+
+    // These dummies result in a tremendous performance improvement, ~300%+
+    _dummy_1: CachePadded<u64>,
+    // data for the consumer
+    head: AtomicUsize,
+    tail_cache: AtomicUsize,
+    prod_alive:AtomicBool, //seems weird, but consumer will read this
+
+    _dummy_2: CachePadded<u64>,
+    // data for the producer
+    tail: AtomicUsize,
+    head_cache: AtomicUsize,
+    cons_alive: AtomicBool, //seems weird, but producer will read this
+}
+
+unsafe impl<T: Send> Send for SpscBufferQueue<T> {}
+
+impl<T: Send> SpscBufferQueue<T> {
+    pub fn new(size: usize) -> (BufferProducer<T>, BufferConsumer<T>) {
+        let q = SpscBufferQueue {
+            data_block: unsafe{ allocate(size) },
+            size: cmp::min(size, (isize::max_value() - 1) as usize) + 1,
+            _marker: PhantomData,
+
+            _dummy_1: CachePadded::zeroed(),
+            head: AtomicUsize::new(0),
+            tail_cache: AtomicUsize::new(0),
+            prod_alive: AtomicBool::new(true),
+
+            _dummy_2: CachePadded::zeroed(),
+            tail: AtomicUsize::new(0),
+            head_cache: AtomicUsize::new(0),
+            cons_alive: AtomicBool::new(true),
+        };
+        let qarc = Arc::new(q);
+        let rtuple = (BufferProducer::new(qarc.clone()),
+                      BufferConsumer::new(qarc));
+        rtuple
+    }
+
+    /// Tries constructing the element and inserts into the queue
+    ///
+    /// Returns the closure if there isn't space
+    #[inline(always)]
+    fn try_construct<F>(&self, ctor: F) -> Result<(), F> where F: FnOnce() -> T {
```

Comment:
I suspect the use of `&self` here is problematic, as it allows calling `try_construct` *inside* the call to `ctor`, which will probably not behave like one expects. Specifically,

```rust
let prod: BufferProducer<i32> = ...;
prod.try_construct(|| {
     prod.try_construct(|| 1);
     2
})
```

I suspect, since this is spsc, all these methods should take `&mut self`, to express the "single" requirement statically and also defend against this problem.